src/prefect/events/clients.py

```diff
@@ -195,3 +214,166 @@ async def _emit(self, event: Event) -> None:
+                print(event.occurred, event.resource.id, event.event)
+
+    """
+
+    _websocket: Optional[WebSocketClientProtocol]
+    _filter: "EventFilter"
+    _seen_events: "TTLCache[UUID, bool]"
+
+    def __init__(
+        self,
+        api_url: str = None,
+        api_key: str = None,
+        filter: "EventFilter" = None,
+        reconnection_attempts: int = 10,
+    ):
+        """
+        Args:
+            api_url: The base URL for a Prefect Cloud workspace
+            api_key: The API of an actor with the manage_events scope
+            reconnection_attempts: When the client is disconnected, how many times
+                the client should attempt to reconnect
+        """
+        api_url = api_url or PREFECT_API_URL.value()
+        api_key = api_key or PREFECT_API_KEY.value()
+
+        from prefect.events.filters import EventFilter
+
+        self._filter = filter or EventFilter()
+
+        self._seen_events = TTLCache(maxsize=500_000, ttl=120)
+
+        socket_url = (
+            api_url.replace("https://", "wss://")
+            .replace("http://", "ws://")
+            .rstrip("/")
+        ) + "/events/out"
+
+        logger.debug("Connecting to %s", socket_url)
+
+        self._api_key = api_key
+        self._connect = connect(
+            socket_url,
+            subprotocols=["prefect"],
+        )
+        self._websocket = None
+        self._reconnection_attempts = reconnection_attempts
+
+    async def __aenter__(self) -> "PrefectCloudEventSubscriber":
+        # Don't handle any errors in the initial connection, because these are most
+        # likely a permission or configuration issue that should propagate
+        await self._reconnect()
+        return self
+
+    async def _reconnect(self) -> None:
+        logger.debug("Reconnecting...")
+        if self._websocket:
+            self._websocket = None
+            await self._connect.__aexit__(None, None, None)
+
+        self._websocket = await self._connect.__aenter__()
+
+        # make sure we have actually connected
+        logger.debug("  pinging...")
+        pong = await self._websocket.ping()
+        await pong
+
+        logger.debug("  authenticating...")
+        await self._websocket.send(
+            orjson.dumps({"type": "auth", "token": self._api_key}).decode()
+        )
+
+        try:
+            message = orjson.loads(await self._websocket.recv())
+            logger.debug("  auth result %s", message)
+            assert message["type"] == "auth_success"
+        except (AssertionError, ConnectionClosedError) as e:
+            if isinstance(e, AssertionError) or e.code == WS_1008_POLICY_VIOLATION:
+                raise Exception(
+                    "Unable to authenticate to the event stream. Please ensure the "
+                    "provided api_key you are using is valid for this environment."
+                ) from e
+            raise
+
+        from prefect.events.filters import EventOccurredFilter
```

Comment:
just curious, why are we deferring this import?

Reply:
Good eye, the `prefect.events.client` package is imported quite early in the chain of prefect imports, because `emit_event` is used by the engine.  Unless these are deferred, we'll end up with a circular reference involving `PrefectBaseModel` and `prefect.states`.  I would love to take a pass through and fix up some of the eager imports to calm things down at import time.

Reply:
gotcha, makes sense - thanks!